### PR TITLE
chore(docs): refactor payments frontend

### PR DESCRIPTION
**Pull request by StaberindeZA**
## Because

- The subscription platform wants to refactor the payments frontend.

## This pull request

- Adds the ADR discussing various payments frontend refactor solutions.

## Issue that this pull request solves

Closes: #FXA-7014

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: docs/adr/0035-refactor-payments-frontend.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+- Easier routing - The SubPlat teams wants to perform more advanced routing to allow features such as, Relying Party specific checkout pages, A/B testing, etc,
+- GraphQL support - Improve app performance and responsivness by reducing the number of network requests
+- Reuse existing components - We have received generally favourable reviews of our existing frontend look and feel, and want to keep the appearance the same
+
+## Considered Options
+
+- A. React Framework - Next.js
+- B. React Framework - Gatsby.js
+- C. React, Typescript, without Redux
+- D. React, Typescript, with Redux
```
Comment (chenba) [inline docs/adr/0035-refactor-payments-frontend.md:29]: Remix is the new hotness I think.  Would you consider it?

File: docs/adr/0035-refactor-payments-frontend.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+- Bad, because it's new to most team members, and requires some learning to use more complicated features
+- Bad, because it diverges from tooling used elsewhere in the FxA monorepo
+
+### C. React, Typescript, without Redux
+
+Use the existing fxa-payments-server app and only remove redux.
+
+- Good, because simplifies the app by removing complexity of using Redux
+- Good, because it would mostly use the existing technologies, and not require new learnings for the team
+- Bad, because Rescripts and other tooling make it difficult to upgrade React tooling. (See [FXA-5242](https://mozilla-hub.atlassian.net/browse/FXA-5242))
```
Comment (chenba) [inline docs/adr/0035-refactor-payments-frontend.md:76]: rescripts was added to add or overwrite configuration from create-react-app.  CRA's the real culprit 😀

File: docs/adr/0035-refactor-payments-frontend.md (hunk 1 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:29]: Oh interesting, the team actually hadn't considered Remix. I'll bring it up in our team meeting later today. Great suggestion thank you.

Here are some of my initial thoughts.
- Looks really slick. Especially because of this comment from Reddit 
  - "...[my mind was blown away by how smooth and relaxing my development experience became](https://www.reddit.com/r/reactjs/comments/112bujy/switched_from_nextjs_to_remixjs_and_loving_it/)..."
- We are currently probably looking at an SSR app, and if performance is our main driver, the comment below makes me think maybe there won't be as big a performance difference?
  - "[If the Next.js app moved away from client fetching, and used getServerSideProps, they would probably close the gap and have simpler answers to these questions.](https://remix.run/blog/remix-vs-next)..."
- The [drawbacks mentioned here](https://blog.logrocket.com/guide-to-remix-react-framework/#drawbacks-remix) as well as the [conclusion from this piece](https://dev.to/braincuber_technologies/remix-vs-nextjs-a-detailed-comparison-2ogg), are probably my main concerns.
- Do you happen to know if Remix is used by any other teams in Mozilla? I know Next.js is.

Comment (chenba) [reply docs/adr/0035-refactor-payments-frontend.md:29]: > Do you happen to know if Remix is used by any other teams in Mozilla? I know Next.js is.

No.  I don't know what other teams use.

Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:29]: Just to circle back to this. We discussed this as a team, and decided to stick with Next. Along with the drawbacks mentioned above, we primarily favor Next for its popularity, community size, proven production use, and that its already used by other teams at Mozilla.

Good suggestion though, and I'll update the docs with the option.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
+# Refactor Subscription Platform frontend
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:1]: Nit: might add "with NestJS" in the title and MD file name to match our others (and so the primary decision is obvious when looking at the ADR file tree)

File: docs/adr/0035-refactor-payments-frontend.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+Additionally, as the app has grown, so have the number of network calls causing a reduction in app performance. The aim is to refactor the app to have one network call for each action.
+
+Finally, the frontend includes some business and application logic, which has made components and parts of the app unnecessarily complex. As a result the frontend is less flexible, and harder to change and maintain. To solve this, the Subscription Platform team wants to move all business and application logic to the backend.
+
+To address the problems briefly discussed above, and to meet current and future relying party and feature requests, refactoring of the existing Subscription Platform is necessary.
+
+## Decision Drivers
+
+- Faster development - This applies to both the refactoring necessary for SubPlat 3.0, as well as future relying party and feature requests
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:19]: Might make a note about what "SubPlat 3.0" is since reading this out of context might be confusing, even if it's just "dubbed SubPlat 3.0" at the end of your last sentence under the context & problem statement.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+- Easier routing - The SubPlat teams wants to perform more advanced routing to allow features such as, Relying Party specific checkout pages, A/B testing, etc,
+- GraphQL support - Improve app performance and responsivness by reducing the number of network requests
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:21]: Using GQL also promotes consistency with other parts of the FxA stack.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+Use a React framework, Next.js.
+
+- Good, because it provides many benefits out of the box, without compromising flexibility, that would require additional libraries otherwise. (Routing, rendering options, image optimization, etc.)
+- Good, because of its routing simplicity and flexibility
+- Good, because of its built in tooling, which reduces existing bespoke tooling currently used
+- Good, because it is easy to learn the basics
+- Good, because it's parent company has other tools that work well with Next.js and enhance its capabilities. (Turbopack, SWR, etc.)
+- Good, because of its large community support and popularity
+- Good, because it is used by other teams in Mozilla
+- Good, because it no longer uses Redux for state management
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:51]: maybe add, "which we have determined is overly complex for our needs". What is the state management plan, is it Apollo Cache + local hooks + Context?

File: docs/adr/0035-refactor-payments-frontend.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+- Bad, because it adds another layer on top of React
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:52]: And another dependency.

Also, are we going to use most of the features NextJS offers? It could be a con that it's heavier than vanilla React and we may not utilize all of its features, but you would know better than me.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+Regarding why Next.js is favored over Gatsby.js, before completing the ADR, the subscription platform team went through a prototyping project and found the two frameworks to be remarkably similar in capabilities. However due to Next.js's popularity, use by other teams in Mozilla and a team vote, the decision was made to pick Next.js.
+
+## Pros and Cons of the Options
+
+### A. React Framework - Next.js
+
+Use a React framework, Next.js.
+
+- Good, because it provides many benefits out of the box, without compromising flexibility, that would require additional libraries otherwise. (Routing, rendering options, image optimization, etc.)
+- Good, because of its routing simplicity and flexibility
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:45]: It's mentioned a couple of times that the routing is nicer and more flexible and even nicer for RPs. It could be worth noting what it offers specifically besides abstraction into a `pages/` directory?

It could also be a neutral point that NextJS is more opinionated than vanilla React. Nice for some things, might could limit us other times.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+Regarding why Next.js is favored over Gatsby.js, before completing the ADR, the subscription platform team went through a prototyping project and found the two frameworks to be remarkably similar in capabilities. However due to Next.js's popularity, use by other teams in Mozilla and a team vote, the decision was made to pick Next.js.
+
+## Pros and Cons of the Options
+
+### A. React Framework - Next.js
+
+Use a React framework, Next.js.
+
+- Good, because it provides many benefits out of the box, without compromising flexibility, that would require additional libraries otherwise. (Routing, rendering options, image optimization, etc.)
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:44]: We probably want to mention server-side rendering here right?

It could be worth referencing [this ADR](https://github.com/mozilla/fxa/blob/main/docs/adr/0010-transition-fxa-from-backbone-to-react.md#option-c---move-to-isomorphic-react) for some pros or even that explanation up top of what SSR is. Also probably worth linking to it and explaining what we've learned since then (though granted that ADR was for fxa settings and not subplat).

File: docs/adr/0035-refactor-payments-frontend.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+- Bad, because it's new to most team members, and requires some learning to use more complicated features
+- Bad, because it diverges from tooling used elsewhere in the FxA monorepo
+
+### C. React, Typescript, without Redux
+
+Use the existing fxa-payments-server app and only remove redux.
+
+- Good, because simplifies the app by removing complexity of using Redux
+- Good, because it would mostly use the existing technologies, and not require new learnings for the team
+- Bad, because Rescripts, CRA, and other tooling make it difficult to upgrade React tooling. (See [FXA-5242](https://mozilla-hub.atlassian.net/browse/FXA-5242))
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:77]: I really only just looked into this yesterday but another option is Vite + Vite's SSR plugin. [See a comparison](https://github.com/brillout/vite-plugin-ssr/issues/158#issuecomment-1250068578) with Next. Not saying that should be in this ADR since it's already a couple weeks along but, I imagine FxA is going to be moving away from CRA sometime. We used it originally because at the time, it really was one of the best options for reasons laid out in another ADR. However now, we could either use Vite instead of CRA, or wait and see what SubPlat thinks of Next 😏

File: docs/adr/0035-refactor-payments-frontend.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+- Good, because of its routing simplicity and flexibility
+- Good, because of its built in tooling, which reduces existing bespoke tooling currently used
+- Good, because it is easy to learn the basics
+- Good, because it's parent company has other tools that work well with Next.js and enhance its capabilities. (Turbopack, SWR, etc.)
+- Good, because of its large community support and popularity
+- Good, because it is used by other teams in Mozilla
+- Good, because it no longer uses Redux for state management
+- Bad, because it adds another layer on top of React
+- Bad, because it's new to most team members, and requires some learning to use more complicated features
+- Bad, because it diverges from tooling used elsewhere in the FxA monorepo
```
Comment (LZoog) [inline docs/adr/0035-refactor-payments-frontend.md:54]: Another con could be level of effort and time to reach production with the refactor since the stack, it sounds like you're basically going to have to rewrite most of the app.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 5 above)
Comment (chenba) [reply docs/adr/0035-refactor-payments-frontend.md:21]: Is GQL support tied into a frontend framework somehow?

File: docs/adr/0035-refactor-payments-frontend.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+- Bad, because it will still require refactoring to introduce GraphQL, remove application logic, and reduce component complexity
+- Bad, because Rescripts, CRA, and other tooling make it difficult to upgrade React tooling. (See [FXA-5242](https://mozilla-hub.atlassian.net/browse/FXA-5242))
+- Bad, because page routing with react-router is complicated
+
+### E. React Framework - Remix
+
+Use the existing Subscription Platform, and only refactor existing code.
+
+- Good, because it provides an improved developer experience
+- Good, because it provides great page performance
```
Comment (biancadanforth) [inline docs/adr/0035-refactor-payments-frontend.md:96]: I don't see page performance mentioned for the other options. What's good about the page performance here? Is it better than the other options in some way?

IIUC app performance isn't a major concern/decision driver, though we will likely see some improvements as a side effect of making fewer network requests and moving more logic/state management to the back end.

**Review by biancadanforth — COMMENTED** [review]
Added some comments mostly with some questions. Can we link to the prototypes and/or team discussions somewhere in the ADR? I think you did that for the hCMS ADR.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 12 above)
Comment (bbangert) [reply docs/adr/0035-refactor-payments-frontend.md:96]: Time to interact should definitely be considered for any page in a flow we want users to complete as there's pretty decent quantities of studies at this point about falloffs from delay, e.g.
https://fabric.inc/blog/time-to-interact/

File: docs/adr/0035-refactor-payments-frontend.md (hunk 6 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:51]: The aim is to greatly simplify and only use Context.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+- Neutral, because it only provides server side rendering
```
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:97]: Next.js and Gatsby.js provide other rendering options, like static site generation, and client-side rendering and give you the option of using either one on a page-by-page basis.

Therefore, even though our plan at the moment is to primarily use SSR, some pages we might want to be Static or SSG, or even completely client-side. (Thinking of error pages, or a product/plan catalogue potentially)

Remix only allows for SSR and does not provide the flexibility to use other rendering options, which is how I landed on Neutral, as opposed to bad.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+- Bad, because it is not as production tested and mature a project as Next or Gatsby
+- Bad, because of potentially confusing routing
```
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:99]: I've changed this to Neutral and reworded it slightly. One of the features of Remix is that it allows for nested routes, and [a review](https://blog.logrocket.com/guide-to-remix-react-framework/#drawbacks-remix) mentioned that it could potentially be confusing depending on how it's used.

```
Neutral, because of confusing nested routing functionality
```

File: docs/adr/0035-refactor-payments-frontend.md (hunk 12 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:96]: I explicitly mentioned it here, since the Remix material claims greater performance over competitors like Next. That said, In a discussion during our Engineering Weekly, we concluded that the performance gains are marginal and that Next is nearly as performant.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 9 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:44]: At the moment SSR is indirectly mentioned by `rendering options`.

SSR wasn't really a decision driver for this ADR. It's more that the framework offers a lot out of the box, which allows for faster development.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+- Bad, because it will still require significant refactoring to remove Redux and introduce GraphQL
+- Bad, because page routing with react-router is complicated
+
+### D. React, Typescript, with Redux
+
+Use the existing Subscription Platform, and only refactor existing code.
+
+- Good, because it requires the least amount of new technologies and libraries
+- Bad, because staying with redux keeps complexity in the app without much benefit
+- Bad, because it will still require refactoring to introduce GraphQL, remove application logic, and reduce component complexity
```
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:87]: There was a related comment earlier in this PR, and I've updated to include the following line for the framework options.

```
Bad, because it requries significant refactoring, starting from a clean slate, simplyfing and reusing existing components and styling where possible.
```

File: docs/adr/0035-refactor-payments-frontend.md (hunk 5 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:21]: In Gatsby's case it has a GraphQL data layer built in.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 8 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:45]: I've updated the point as follows. I'd like to keep it in the Good column, since in this case it hits 2 of the decision drivers, Faster development and Easier routing.

```
Good, because of its routing simplicity and flexibility. The file-based routing is simple to understand as a developer, removes some boilerplate code, and easily allows changes to available routes. Although not as flexible as other libraries, it still provides the dynamic routes required for this refactor.
```

File: docs/adr/0035-refactor-payments-frontend.md (hunk 11 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:54]: Ah I forgot to mention that I added the following line for this.

```Bad, because it requries significant refactoring, starting from a clean slate, simplyfing and reusing existing components and styling where possible.```

File: docs/adr/0035-refactor-payments-frontend.md (hunk 9 above)
Comment (LZoog) [reply docs/adr/0035-refactor-payments-frontend.md:44]: Without SSR, wouldn't moving logic to the server (mentioned here) be more difficult since it'd require the client to spin up and then send requests out vs sending data back with the first render? The flexibility at least is nice.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 8 above)
Comment (LZoog) [reply docs/adr/0035-refactor-payments-frontend.md]: 👍 I think it's fine to leave the routing in the good column, my 2nd point was about NextJS being more opinionated than vanilla React (which we may not like in some scenarios), but maybe that's also all fine and we like that.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+### C. React, Typescript, without Redux
+
+Use the existing fxa-payments-server app and only remove redux.
+
+- Good, because simplifies the app by removing complexity of using Redux
+- Good, because it would mostly use the existing technologies, and not require new learnings for the team
+- Bad, because Rescripts, CRA, and other tooling make it difficult to upgrade React tooling. (See [FXA-5242](https://mozilla-hub.atlassian.net/browse/FXA-5242))
+- Bad, because it will still require significant refactoring to remove Redux and introduce GraphQL
+- Bad, because page routing with react-router is complicated
```
[the remark this answers is not in the document]
Comment (LZoog) [reply docs/adr/0035-refactor-payments-frontend.md:81]: I personally don't find react-router that complex, but maybe SubPlat's experience was different?

File: docs/adr/0035-refactor-payments-frontend.md (hunk 9 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:44]: Oh I think I see where the mix up might be happening. When the ADR references `...move all business and application logic to the backend...`, it doesn't necessarily mean the Next.js server, but also the payments APIs in `fxa-auth`.

A good example of this is the logic in the `'/oauth/mozilla-subscriptions/customer/plan-eligibility/{planId}'` API in `fxa-auth`, which used to be done in React.

The reason why this is mentioned in the ADR is to try illustrate that the frontend refactor will change/simplify quite a bit of the frontend, making it an ideal time to adopt a framework. It wasn't necessarily added to indicate that we want a SSR solution. (All that said, at the moment it does look like we'll be SSR first for our pages, and might have some exceptions)

File: docs/adr/0035-refactor-payments-frontend.md (hunk 8 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md]: Yup, I think in this case we do like that it's more opinionated. Although I did find it to still be flexible and still allows the features we are looking for.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 10 above)
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:79]: Thank you for mentioning it! I had not heard of Vite before. Hopefully we'll be able to report back on Next.js in the near future! :D

File: docs/adr/0035-refactor-payments-frontend.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:81]: That's valid, that's been my experience with `react-router` as well. In this case it is just another library for folks to familiarize themselves with vs just copy-pasting a file. Which is something we are thinking of potentially doing, by allowing RPs to copy a template page and configure it themselves.

File: docs/adr/0035-refactor-payments-frontend.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
+# Refactor Subscription Platform frontend
+
+- Status: proposed
+- Deciders: Ben Bangert, Lisa Chan, Barry Chen
```
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:4]: Oh OK, I guess I'll remove it then. 😞 One tries to do something nice... 😉

File: docs/adr/0035-refactor-payments-frontend.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (biancadanforth) [reply docs/adr/0035-refactor-payments-frontend.md:81]: > Would it be OK to reword this to the following?
> 
> `Bad, because page routing with react-router is more complicated, needing to learn a routing library, compared to the file-based routing used by popular React frameworks.`

Thanks for clarifying -- can you link to `react-router` and an example of this "file-based routing" as part of the sentence? That'd probably help folks like me understand the distinction a bit better.

Ah you know I just read up a bit on `react-router`, as it's been a while. It seems like something we'd use if we were keeping the SPA approach on the front-end, but it sounds like we want to go with a multi-page approach instead with actual network requests in between and browser navigation instead of updating the browsing history and URL to get some of the same browser features (i.e. clicking Back and Forward). Maybe you can reference something about that in the sentence too?

File: docs/adr/0035-refactor-payments-frontend.md (hunk 12 above)
Comment (biancadanforth) [reply docs/adr/0035-refactor-payments-frontend.md:98]: So is this worth mentioning then?

Comment (StaberindeZA) [reply docs/adr/0035-refactor-payments-frontend.md:98]: I'm actually thinking of doing the opposite and just removing this line. TTI is definitely important, but each of the options considered in this ADR have some means to improve on TTI to bring it within a range that's acceptable.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 9 above)
Comment (LZoog) [reply docs/adr/0035-refactor-payments-frontend.md]: Ah, that makes more sense. Maybe that's good to clarify then - when I think of NextJS, the two things that immediately come to mind are "React framework" and "SSR" (thirdly SSG). I suppose IMO if I was looking at this ADR in the future I might want to see what we thought about SSR at the time and might make assumptions reading that some logic would be moved to the back-end that it would mean the Next back-end, as I'm guessing there's at least a couple scenarios where sending down data on `getInitialProps` or `getServerSideProps` would be beneficial. I know it's not the primary reason for the ADR so I'll leave it up to you on what's worth noting vs not noting. Maybe I'm just projecting since if/when FxA uses Next we'll definitely have a note about SSR in an ADR since there's been a couple places I've seen during our Backbone to React refactoring work where it would help us out.

File: docs/adr/0035-refactor-payments-frontend.md (hunk 6 above)
Comment (LZoog) [reply docs/adr/0035-refactor-payments-frontend.md]: Apollo Client (sorry I said Apollo Cache heh) would give you some caching features and other data management options like pagination and an optimistic response update (which granted we're not really using in FxA), however, it feels fine to me personally to start off with just Context to reduce complexity. If you find you're needing to store GQL query/responses on Context or you're updating data that updates semi-frequently, you might consider using it then.

**Review by LZoog — APPROVED** [review]
Not sure if it's my place to r+ here but I'mma do it since I've commented back and forth a few times. 👍

**Review by bbangert — APPROVED** [review]
